crates/next-core/src/next_client_reference/visit_client_reference.rs

```diff
@@ -148,14 +197,68 @@ pub async fn client_reference_graph(
             client_references_by_server_component,
             server_component_entries,
             server_utils,
+            visited_nodes: VisitedClientReferenceGraphNodes(visited_nodes.0).cell(),
         }
         .cell())
     }
     .instrument(tracing::info_span!("find client references"))
     .await
 }
 
-struct VisitClientReference;
+#[turbo_tasks::value(shared)]
+#[derive(Clone, Debug)]
+pub struct ServerEntries {
+    pub server_component_entries: Vec<Vc<NextServerComponentModule>>,
+    pub server_utils: Vec<Vc<Box<dyn Module>>>,
+}
+
+#[turbo_tasks::function]
+pub async fn find_server_entries(entry: Vc<Box<dyn Module>>) -> Result<Vc<ServerEntries>> {
+    let graph = AdjacencyMap::new()
+        .skip_duplicates()
+        .visit(
+            vec![VisitClientReferenceNode {
+                state: {
+                    VisitClientReferenceNodeState::Entry {
+                        entry_path: entry.ident().path().resolve().await?,
+                    }
+                },
+                ty: VisitClientReferenceNodeType::Internal(entry, entry.ident().to_string().await?),
+            }],
+            VisitClientReference {
+                stop_at_server_entries: true,
+            },
+        )
+        .await
+        .completed()?
+        .into_inner();
+
+    let mut server_component_entries = vec![];
+    let mut server_utils = vec![];
+    for node in graph.reverse_topological() {
+        match &node.ty {
+            VisitClientReferenceNodeType::ServerUtilEntry(server_util, _) => {
+                server_utils.push(*server_util);
+            }
+            VisitClientReferenceNodeType::ServerComponentEntry(server_component, _) => {
+                server_component_entries.push(*server_component);
+            }
+            VisitClientReferenceNodeType::Internal(_, _)
+            | VisitClientReferenceNodeType::ClientReference(_, _) => {}
+        }
+    }
+
+    Ok(ServerEntries {
+        server_component_entries,
+        server_utils,
+    }
+    .cell())
+}
+
+struct VisitClientReference {
+    /// Used to discover ServerComponents and ServerUtils
+    stop_at_server_entries: bool,
+}
```

Comment:
I feel like it would be better to split that into two different implementations: `VisitFindClientReference` and `VisitFindServerEntries`. Currently we have a lot of unnecessary logic to cover both cases at the same time.

`VisitFindServerEntries` would not need to cast for `EcmascriptClientReferenceModule` and `CssModuleAsset` since we are only looking for server entries.

`VisitFindClientReference` would not need to track the `state` of the current server entry if it is only invoked for a single server entry. The return value could also be simpler: `pub client_references: Vec<ClientReferenceType>, pub visited_nodes: Vc<VisitedClientReferenceGraphNodes>` would be enough.

Only the merging would lead to map of client references per server entry.